### PR TITLE
[otbn/doc] Secure Wipe documentation update.

### PR DESCRIPTION
**Pull request by vrozic**
Added specification of the secure wipe mechanism.
Modified otbn.hjson to include register interface.

Signed-off-by: Vladimir Rozic <vrozic@lowrisc.org>

### CODE REVIEW

File: hw/ip/otbn/doc/_index.md
Comment (rswarbrick) [inline hw/ip/otbn/doc/_index.md:424]: I'm not sure this is quite the right phrasing. Maybe just "This wiping is split into three parts"?

File: hw/ip/otbn/doc/_index.md
Comment (rswarbrick) [inline hw/ip/otbn/doc/_index.md:427]: Could these items link to the relevant sections? Also, are the capitals really necessary? It sounds a bit shouty to me.

File: hw/ip/otbn/doc/_index.md
Comment (rswarbrick) [inline hw/ip/otbn/doc/_index.md:429]: Please add a link to the relevant register for the bus interface.

File: hw/ip/otbn/doc/_index.md
Comment (rswarbrick) [inline hw/ip/otbn/doc/_index.md:431]: I think this is just the internal state secure wipe, right?

File: hw/ip/otbn/doc/_index.md
Comment (rswarbrick) [inline hw/ip/otbn/doc/_index.md:625]: This link is to the wrong section: it should point at the DMEM secure wipe, I think.

File: hw/ip/otbn/doc/_index.md
Comment (rswarbrick) [inline hw/ip/otbn/doc/_index.md:641]: Again, wrong section: should be the IMEM secure wipe, I think.

File: hw/ip/otbn/doc/_index.md
Comment (rswarbrick) [inline hw/ip/otbn/doc/_index.md:656]: Could we rename this and the IMEM one to something like "DMEM key rotation"? I think it would make the other text a bit clearer, avoiding repeating various types of "wipe", which gets a bit confusing.

File: hw/ip/otbn/doc/_index.md
Comment (rswarbrick) [inline hw/ip/otbn/doc/_index.md:662]: The "deterministic in time" phrasing is a bit clunky and is redundant, given that this takes a single cycle. Maybe just "This takes a single cycle."?

File: hw/ip/otbn/doc/_index.md
Comment (rswarbrick) [inline hw/ip/otbn/doc/_index.md:674]: Again, maybe just "This takes a single cycle."?

File: hw/ip/otbn/doc/_index.md
Comment (rswarbrick) [inline hw/ip/otbn/doc/_index.md:666]: I think you mean "data memory secure wipe" here? But, as I said above, I think this would be much clearer if we called it something like "DMEM key rotation".

File: hw/ip/otbn/doc/_index.md
Comment (rswarbrick) [inline hw/ip/otbn/doc/_index.md:678]: Wrong name here.

**Review by rswarbrick — COMMENTED** [review]
As well as the specific comments I've left, please could you add links to all the registers that you refer to in the text? E.g. `SEC_WIPE.DMEM`.

File: hw/ip/otbn/doc/_index.md
Comment (imphil) [reply hw/ip/otbn/doc/_index.md:656]: Hm, I'm not sure how much we want to mix implementation details (key rotation) with what it achieves (get rid of data).

Comment (rswarbrick) [reply hw/ip/otbn/doc/_index.md:656]: Hmm, maybe. The problem is that I got quite confused when reading through this patch and I'm supposed to know what's going on already! I think it would really help if the names were more distinct.

File: hw/ip/otbn/doc/_index.md
Comment (vrozic) [reply hw/ip/otbn/doc/_index.md]: corrected in the latest version

File: hw/ip/otbn/doc/_index.md
Comment (vrozic) [reply hw/ip/otbn/doc/_index.md]: corrected

File: hw/ip/otbn/doc/_index.md
Comment (vrozic) [reply hw/ip/otbn/doc/_index.md]: corrected

File: hw/ip/otbn/doc/_index.md
Comment (imphil) [inline hw/ip/otbn/doc/_index.md:431]: This is effectively all one paragraph explaining the trigger mechanisms. Maybe make that clearer with text like this?

```suggestion
A secure wipe is performed automatically in certain situations, or can be requested manually by the host software.
A secure wipe of all state is performed when a fatal error is detected. 
(TODO: Specify lifecycle state dependency.)
A secure wipe of the internal state only is performed whenever an OTBN operation finishes.
Finally, host software can manually trigger a secure wipe operation by writing to the {{< regref "SEC_WIPE">}} register.
```

(Leave the TODO in there for now, it's something we need to figure out later.)

File: hw/ip/otbn/doc/_index.md
Comment (imphil) [inline hw/ip/otbn/doc/_index.md:648]: Can you add a one-line intro sentence to secure wipe here?

File: hw/ip/otbn/doc/_index.md
Comment (imphil) [inline hw/ip/otbn/doc/_index.md:654]: Three comments:
* Can you put this content into a separate section, e.g. "Triggering secure wipe"
* Please avoid non-ASCII characters.
* Define "full secure wipe" somewhere.

File: hw/ip/otbn/doc/_index.md
Comment (imphil) [inline hw/ip/otbn/doc/_index.md:689]:
```suggestion
The wiping procedure is a two-step process:
```

File: hw/ip/otbn/doc/_index.md
Comment (imphil) [inline hw/ip/otbn/doc/_index.md:687]: Many of these registers are actually mappings of each other and point to the same data. So instead of using the register names, maybe something like this would work better:

* Flags (accessible through the FG0, FG1, and FLAGS CSRs)
* The modulus (accessible through the MOD0 to MOD7 CSRs and the MOD WSR)

File: hw/ip/otbn/data/otbn.hjson
Comment (imphil) [inline hw/ip/otbn/data/otbn.hjson:280]:
```suggestion
        Trigger a secure wipe operation.
```

File: hw/ip/otbn/data/otbn.hjson
Comment (imphil) [inline hw/ip/otbn/data/otbn.hjson:288]: "Write 1 to securely wipe the data memory. Always reads as 0."

**Review by imphil — COMMENTED** [review]
Thanks @vrozic, this looks good!

Two questions came to my mind that need clarification:

1. How does OTBN behave while a secure wipe operation is in progress, and how do we signal completion to the host CPU?

   Options I could think about:
   1. Use the done/busy infrastructure we already have, i.e. set busy == 1 during the wipe operation, and issue a done interrupt once that has completed.
   2. Introduce a flag similar to the busy one, e.g. BUSY_SEC_WIPE, and potentially an associated  interrupt, for the secure wipe functionality.
   3. Don't signal completion and just block access to OTBN.

   Option (i) or (ii) are effectively the "a CSR should be added that allows software to spin wait while a key request is pending." that was mentioned in https://github.com/lowRISC/opentitan/issues/7021#issuecomment-870085210. 
   I'd personally just go for option (i), I can't see a reason why software would want to differentiate between the two "busy" states. After all, it triggered one or the other operation explicitly and knows what it is waiting for.
2. All of OTBN's operations are run-to-completion, with all outside access being blocked until the OTBN software itself decides that it is done (by executing ECALL). That means that none of the secure wipe functionality takes effect as long as OTBN is BUSY. This fact needs to be stated in the register interface.
   We also need to document what happens if host software still tries to initiate a secure wipe while OTBN is busy. Should we ignore the request? Or raise an alert? Can you maybe come up with something that is consistent with how we seal off DMEM/IMEM access and also apply it to other registers, such as START_ADDR (which probably should follow the same rules).

File: hw/ip/otbn/doc/_index.md
Comment (imphil) [reply hw/ip/otbn/doc/_index.md:431]: Thinking further: we certainly want to do a internal secure wipe on a recoverable error (since that's identical to just terminating an OTBN operation early). We might even want to do a full secure wipe in that case. The list of errors which cause a recoverable error is at https://docs.opentitan.org/hw/ip/otbn/doc/index.html#Reg_err_bits and they could be both a programmer error, or a side-effect of a fault injection. 

The answer probably brings us back to the debuggability vs. security discussion.

@vrozic can you update the document to always include internal wipe for recoverable alerts, and leave a TODO for the potential full secure wipe? I could imagine the outcome being "full secure wipe if in PROD mode, otherwise internal wipe only."

File: hw/ip/otbn/doc/_index.md
Comment (imphil) [inline hw/ip/otbn/doc/_index.md:442]: nit:

```suggestion
A secure wipe of only the internal state is performed whenever an OTBN operation is completed and after a recoverable error.
```

File: hw/ip/otbn/doc/_index.md
Comment (imphil) [inline hw/ip/otbn/doc/_index.md:445]: The section this text appears in is the high-level description of security features. I would argue that the busy/interrupt behavior is better placed in the design details section further down.

And a nit:
```suggestion
Once the operation is completed, a {{< regref "INTR_STATE.done" >}} interrupt is raised and {{< regref "STATUS.busy">}} is cleared. 
```

On that topic: It is already implicitly defined, but maybe add one explicit sentence that OTBN being busy during the secure wipe operation, and writes to SEC_WIPE being ignored while OTBN is busy, means effectively that host software will get a single done interrupt for a secure wipe operation, independent of how many SEC_WIPE bits the software wrote.

File: hw/ip/otbn/doc/_index.md
Comment (imphil) [inline hw/ip/otbn/doc/_index.md:447]:
```suggestion
Refer to the [Secure Wipe]({{<relref "#design-details-secure-wipe">}}) section for implementation details.
```

(To match the section title below.)

File: hw/ip/otbn/doc/_index.md
Comment (imphil) [inline hw/ip/otbn/doc/_index.md:679]: We use the term "internal state" in this context as separate category next to instruction and data memories.

```suggestion
The entire OTBN state, including the contents of the instruction and data memories, can be securely deleted on demand from a host software.  
```

File: hw/ip/otbn/doc/_index.md
Comment (imphil) [inline hw/ip/otbn/doc/_index.md:689]: Two small things:
* We are not super consistent, but generally tried to go with "error" as the condition that is observed, and "alert" being the reaction to an error.
* For recoverable errors, we still "complete" the operation, just not successfully.

```suggestion
The internal state secure wipe is automatically triggered when an OTBN operation completes, either successfully, or unsuccessfully due to a recoverable error.
```

File: hw/ip/otbn/doc/_index.md
Comment (imphil) [inline hw/ip/otbn/doc/_index.md:691]:
```suggestion
Host software can trigger a full state wipe by writing 3'b111 to {{< regref "SEC_WIPE">}}  (i.e. set all individual state wipe bits).
```

File: hw/ip/otbn/doc/_index.md
Comment (imphil) [inline hw/ip/otbn/doc/_index.md:703]: Add a reference here to a section describing how completion of the operation is signaled.

File: hw/ip/otbn/doc/_index.md
Comment (imphil) [inline hw/ip/otbn/doc/_index.md:715]: As above

File: hw/ip/otbn/doc/_index.md
Comment (imphil) [inline hw/ip/otbn/doc/_index.md:733]: And as above again.

**Review by imphil — COMMENTED** [review]
Thanks @vrozic! LGTM with a couple small suggestions.

Please configure your editor to remove all trailing spaces at the end of a line.

File: hw/ip/otbn/doc/_index.md
Comment (imphil) [reply hw/ip/otbn/doc/_index.md]: @vrozic did you intentionally not update this part, or did you just overlook it?

**Comment by imphil** [comment]
Thanks for the update @vrozic. 

This looks good now, modulo one comment which might have fallen through the cracks. Thanks for your work on it!

To ensure nothing gets lost between updates, can you please click on the "resolve" button for all comments you've addressed, and write a quick note in the pull request?

@rswarbrick would you mind giving the whole thing a read-through as well before we merge?

**Review by imphil — APPROVED** [review]

File: hw/ip/otbn/data/otbn.hjson
Comment (rswarbrick) [inline hw/ip/otbn/data/otbn.hjson:300]: I don't think we document the expected read value for write-only registers. If you're force-pushing again, could you drop the second sentence for each of these fields?

File: hw/ip/otbn/doc/_index.md
Comment (rswarbrick) [inline hw/ip/otbn/doc/_index.md:691]: Nit: Could you surround the 3'b111 with back-ticks? Otherwise, the apostrophe character will get rendered like normal punctuation, which might look a bit odd.

File: hw/ip/otbn/doc/_index.md
Comment (rswarbrick) [inline hw/ip/otbn/doc/_index.md:710]: Nit: raising a**n** (because INTR_STATE starts with a vowel). Similarly for the next two sections.

**Review by rswarbrick — APPROVED** [review]
A couple of nits, but looks good otherwise.

**Comment by imphil** [comment]
I'm assuming you're done with this @vrozic and we can merge. (Please state so explicitly in the future to avoid misunderstandings!)

**Pull request merged**
